dvc/tree/osf.py

```diff
@@ -0,0 +1,145 @@
+import logging
+import os
+import threading
+
+from funcy import cached_property, wrap_prop
+
+from dvc.exceptions import DvcException
+from dvc.hash_info import HashInfo
+from dvc.path_info import URLInfo
+from dvc.progress import Tqdm
+from dvc.scheme import Schemes
+
+from .base import BaseTree
+
+logger = logging.getLogger(__name__)
+
+
+class OSFAuthError(DvcException):
+    def __init__(self):
+        message = (
+            "OSF authorization failed. Please check or provide"
+            " osf_username and password. See "
+            "https://bit.ly/375On32 for details"
+        )
+        super().__init__(message)
+
+
+class OSFTree(BaseTree):
+    scheme = Schemes.OSF
+    REQUIRES = {"ofsclient": "osfclient"}
+    PARAM_CHECKSUM = "md5"
+    PATH_CLS = URLInfo
+
+    def __init__(self, repo, config):
+        super().__init__(repo, config)
+
+        self.path_info = self.PATH_CLS(config["url"])
+
+        self.osf_username = config.get("username")
+        self.project_guid = config.get("project")
+        self.password = os.getenv(
+            "OSF_PASSWORD", None
+        )  # need for private projects
+        if self.password is None:
+            self.password = config.get("password")
+        logger.debug(OSFTree)
+
+    @wrap_prop(threading.Lock())
+    @cached_property
+    def osf(self):
+        import osfclient
+
+        osf = osfclient.OSF()
+        return osf
+
+    @wrap_prop(threading.Lock())
+    @cached_property
+    def project(self):
+        from osfclient.exceptions import UnauthorizedException
+
+        osf = self.osf
+        osf.login(self.osf_username, self.password)
+        try:
+            project = osf.project(self.project_guid)
+        except UnauthorizedException:
+            raise OSFAuthError
+        return project
+
+    @wrap_prop(threading.Lock())
+    @cached_property
+    def storage(self):
+        return self.project.storage()
+
+    def _get_file_obj(self, path_info):
+        for file in self.storage.files:
+            if file.path == path_info.path:
+                return file
+
+    def get_md5(self, path_info):
+        file = self._get_file_obj(path_info)
+        md5 = file.hashes.get("md5")
+        return md5
+
+    def exists(self, path_info, use_dvcignore=True):
+        paths = self._list_paths()
```

Comment:
this can be extremely inefficient if remote is large. Is there a call that checks a singe file?

DVC most likely expects for remotes to implement `exists` as a single call that check one specific file. Is it possible here?

In general for all remotes we have a pretty good mechanism that given a few settings and "pin-point" `exists` + `walk_files` implementations can decide which specific mechanism to use to run `dvc status`, for example (a lot of exists calls or one single list).

@pmrowla could you take a look and recommend something in this case please?

Reply:
@shcheklein, @pmrowla  every item in OSF storage (folder or file) is a node of tree and store links to a childes. Class ```Storage``` in ```osfclient``` lib provide ```files``` atribute (iterator) to iterate ower these tree.  Is there any way we can determine which folder the file stored in based on the ```path_info``` value? If so, I think it is possible to avoid going through the whole tree here.

Reply:
@aerubanov yes, `path_info` is the regular path pretty much - `a/b/c`, so you do know `a/b` - directories path. Is it enough to check if efficiently?

> Class Storage in osfclient lib provide files atribute (iterator) to iterate ower these tree. 

so, is it lazy? or does it fetch the whole tree from the remote when you first initialize this?

Reply:
Yes, it is lazy. I think that we can firstly iterate over folders, and then iterate over files only in selected folder to find some specific file. It will be faster then iterate over all files in storage.

Reply:
I need to take another look at the `osfclient' code to be 100% sure, but I think I can implement it.